### ISSUE TITLE
float16

### ISSUE THREAD

**Issue by AlexanderMath** [orbital-materials/orb-models#71]
**Question 1.** Did anyone get Orb running in `float16` or `bfloat16`? 

**Question 2.** If I try and e.g. finetune model in `float16`, what would convince us model didn't break? (e.g. validation MAE on MPTraj)

**Comment by AlexanderMath** [comment]
notes on timing orb+rtx4090+[crispr](https://www.rcsb.org/structure/7S4X) (36k atom incl Hs and waters)
- `orb_d3_v2` got 1it/s  [traj](https://rcsb.ai/dd51d28eae?stride=10) 100% nvidia-smi util [1]
- `orb_d3_v2`  got 2.27it/s with bf16+torch.compile [traj](https://rcsb.ai/d969f54d5b) ~80% util [2]
- `orb_d3_xs_v2` got 4.1it/s with bf16+torch.compile [traj](https://rcsb.ai/024e54ee56)  ~50% util [3]

surprises: 
- `bf16` and `torch.compile` only improved when combined  
- `orb_d3_xs_v2` only reduce `message_passing=15->5` but drop GPU util `100%->50%` as  `nvidia-smi`   
- fixing `orb_d3_xs_v2` GPU util could give 8it/s or 0.34ns/day for CRISPR on cheap $0.2/h rtx4090 (or $14/ns). 

todo
[ ] eval on PDB without missing residues. compute `orb_d3_v2` report `MAE(orb_d3_v2, orb_d3_xs_v2_bf16)` for frames from `orb_d3_v2` trajectory
[ ] same eval but ?MPRaj?

[1] 
![Image](https://github.com/user-attachments/assets/28b9fdad-581e-4ab6-9a87-1a51b8f1b4d7)

[2]
![Image](https://github.com/user-attachments/assets/303530c8-7c15-4c6e-a1a1-0342d59cee30)

[3]
![Image](https://github.com/user-attachments/assets/93dda946-c191-45f1-8120-5b3e13629f33)

**Comment by AlexanderMath** [comment]
notes
- low util for `orb_d3_xs_v2` was caused by `brute_force_kNN`;  small hack got me to 6.7it/s and 80% util  for non-pbc (recompute full cdist each 20 iterations=10fs, store 1024 closest atoms;  the 19/20 steps compute only (36000, 1024) distances and pick topk amongst those -- for this to break an atom outside the 1024 closest should move to be top k closest at least 3A within 10fs and i'd expect it to move < 0.1A).

**Comment by AlexanderMath** [comment]
the hack.

```
step = 0
top1024 = None
def brute_force_knn(positions: torch.Tensor, k: int) -> Tuple[torch.Tensor, torch.Tensor]:
    global step, top1024
    if step % 20 == 0: 
        print('recompute all')
        dist = torch.cdist(positions, positions)
        vals, indxs = torch.topk(dist, k, largest=False, sorted=True)
        _ , top1024 = torch.topk(dist, 1024*2, largest=False, sorted=True)
    else: 
        print('recompute topk')
        new_dists = torch.linalg.norm(positions.view(-1, 1, 3) - positions[top1024] , dim=2)
        vals, new_indxs = torch.topk(new_dists, k, largest=False, sorted=True)
        indxs = top1024.gather(1, new_indxs)

    step += 1
    return vals, indxs 
```

warning:  in double precision it passes `th.allclose( old_algo, new_algo )` for my case.  it breaks in two cases 
- in f32 because `cdist` disagrees with `th.linalg.norm(a.view(-1,1,3)-b.view(1,-1,3)`
- if an atom moves enough to go from outside top 1024 to top k; for me this is ~5A in 10fs and guestimate they move <0.1A.

**Comment by AlexanderMath** [comment]
![Image](https://github.com/user-attachments/assets/629e7c37-e340-45a8-b2e0-4d206eaac5fc)

At 6.5step/s = 560k steps/day.  Not sure there's anything left to do wrt model layer inferences. Main thing left is 20ms (~15%) of GPU not doing anything after forward pass (presumably ASE doing it's thing on CPU) -- might take a look, but think this is sufficiently fast if i switch to h100. 

Next
[ ] try timing H100;  best case we get 3x speedup and reach 1.5m steps/day.
[ ] redo [figure 2](https://arxiv.org/pdf/2410.22570) in bf16
[ ] train a bit in bf16 and see how much perf we can recover for figure 2 case.

**Comment by benrhodes26** [comment]
Hi Alexander,

Sorry for the delay in responding; as you can see, we've been busy working on major updates and have just released a new version (0.5.0) with a suite of new models (`orb-v3-*`).

In particular, `orb-v3-direct-20-omat` is the most similar to `orb-v2`, but substantially faster, and more performant in our benchmarking.

Speed-wise, the main changes include: compilation, shallower but wider models (5 layers instead of 15, 1024 width in MLPs rather than 512) and better graph-construction methods (we use cuML KNN routines for large systems).

I would be curious to hear how these changes impact your use-case.

We would also be interested in a PR contribution from you for bf16 --- but please benchmark it against our default `precision=float32-high`, which calls `torch.set_float32_matmul_precision("high")`.

Your hack is also well-motivated, and we're interested in implementing something like that. However, we may take a bit more time to evaluate it in detail, as we don't want to implement something that is flaky and confusing for users.

**Comment by AlexanderMath** [comment]
Christmas arrived early -- amazing results on accuracy/speed trade-off 🎉 

> I would be curious to hear how these changes impact your use-case.

```
# v2
orb_d3_v2 # 1it/s
orb_d3_v2  # 2.27it/s with bf16+torch.compile
orb_d3_xs_v2 # 6.5it/s bf16+compile+kNN hack

# v3
orb-v3-direct-20 # 2.25it/s 
orb-v3-direct-20 # 3.7it/s bf16(gnn_stack) + kNN with cuml=12.2
```

Note:  `gnn_stack` improved from 0.382m to 0.229ms when switching f32 to bf16. I was expecting 3x reduction ~126ms, will debug w/ `nsys profile` when i get back to desktop (`nsys-ui` doesn't work over ssh...).

Note:  rtx4090 is 6x cheaper than a100 (~$0.2/h vs $1.2/h) and has 1.05x more TFLOPs (330 tflops [1] vs 312tflops [2]). 

> We would also be interested in a PR contribution from you for bf16 --- but please benchmark it against our default precision=float32-high, which calls torch.set_float32_matmul_precision("high").

Got it. LMK if you have preference for test-set. 

> Your hack is also well-motivated, and we're interested in implementing something like that. However, we may take a bit more time to evaluate it in detail, as we don't want to risk implementing something that is flaky and confusing for users.

Yeh agree, not sure this would ever be good default behaviour. 

[1] https://images.nvidia.com/aem-dam/Solutions/geforce/ada/nvidia-ada-gpu-architecture.pdf page 30 
[2] https://www.nvidia.com/content/dam/en-zz/Solutions/Data-Center/a100/pdf/nvidia-a100-datasheet-us-nvidia-1758950-r4-web.pdf

**Comment by benrhodes26** [comment]
> Got it. LMK if you have preference for test-set.

We just added `internal/speed.py` (v0.5.1) which can be run like so:
```
python internal/speed.py --method orb-forward --extra_kwargs "precision=float32-high compile=true"
```

**Comment by beckernick** [comment]
Hi! Came across this issue due to the cuML reference. I work on accelerated data science at NVIDIA.

I'm not familiar with this project, but it looks very cool.

I wanted to mention that we've done a lot of work optimizing approximate nearest neighbors algorithms recently. If _exact_ KNN isn't critical, these can provide a major speedup.

We recently migrated cuML's GPU-accelerated UMAP implementation from exact KNN to ANN (using the nn-descent algorithm, which is what the CPU version of UMAP uses). This enabled [huge performance and scalability benefits, without materially affecting UMAP quality](https://developer.nvidia.com/blog/even-faster-and-more-scalable-umap-on-the-gpu-with-rapids-cuml/).

This is currently available in C++ and Python via [cuVS](https://github.com/rapidsai/cuvs), but is something we could explore making available e.g., via cuML's `NearestNeighbors` class, too.

If the KNN is a non-trivial amount of time, could be worth exploring. Happy to chat further, if interested.

**Comment by AlexanderMath** [comment]
update: trying to get the ?10x? cost saving necessary to meet pharmas much lower budget. while h100 has ~500tflops32 at $3-8/h the rtx4090 has 330tflop16 at $0.2/h. so theoretically, in my case, i 1ns should go from ~$120 to $5. looks like i got it working

the hard part was getting the theoretical f16 speedup in practice (rtx4090 has 80tflops32 and 330tflops16). the problem is that in pytorch  `a.to(th.float16)@b.to(th.float16)` doesn't work on rtx4090 because it accumulates in f32. to fix this I had to rewrite the matmul + following SiLU in triton and force the accumulate `tl.dot(a,b,acc=acc, tl.float16)` and then rewrite everything around a triton matmul.
